### PR TITLE
[draft] bindgen poc - v1

### PR DESCRIPTION
**Pull request by jasonish**
See if using bindgen to generate Rust bindings for C functions could have prevented the issues fixed in this PR: https://github.com/OISF/suricata/pull/12009

Ticket: https://redmine.openinfosecfoundation.org/issues/7341

And yes, it would.

- First we define the function pointer type for `SCAppLayerStateGetEventInfoByIdFn`, and some others in `app-layer-ext.h`. Note that I had to bring back the definition of the `AppLayerEventType` enum to C from Rust to avoid circular dependency.
- The header name `app-layer-ext.h` is a bad name, but I just needed a "clean" header.
- Then we integrate C bindgen during build to generate Rust bindings to these functions.
- Replace the hand written bindings with the generated bindings.
  - This happens at commit 3e4f19663e28907e977f53d5d39740c44d6b515e, and the Rust now compiles due to `i8` and `i32` discrepancies.
- Finally, make Rust compile again by fixing the typing errors
- I reverted the fix from #12009 and sure enough the fixes forced on us prevent the ffi issue (that's why this can't be merged according to github)

Caveats:
- As mentioned in the ticket, circular dependencies need to be resolved, a good rule to prevent this is in the ticket.
- bindgen *requires* `libclang`.  Many of the CI jobs will fail because clang is not installed
- This build fresh bindings on every rust compile.  The `bindgen` library requires a lower MSRV than the CLI tool, which requires a newer MSRV that we use right now, and adds another tool dependency. However, we could revert to a style like we do with cbindgen.  Then `bindgen-cli` and `libclang` would not be needed on releases from dist, but dist builders would need newer rust (not bindgen is often found in system packages, so may be a non-issue).

Other notes:
- This will detect mismatched sizes return value integers for C functions exported to Rust
- But it won't catch the case where a different size integer is returned from a Rust function exported C, in the case where Rust might be returning an `i8` and C is storing it in an `int32_t` for instance.

### CODE REVIEW

File: rust/src/sys.rs [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * version 2 along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
+ * 02110-1301, USA.
+ */
+
+#![allow(non_camel_case_types)]
+
```
Comment (jasonish) [inline rust/src/sys.rs:19]: `sys` is used, as naming crates *sometime*-`sys` is idiomatic for crates that pull in a C library into Rust.

**Comment by suricata-qa** [comment]
ERROR:
 

ERROR: QA failed on build_opt.
  

Pipeline 23168

**Comment by suricata-qa** [comment]
Information: QA ran without warnings. 

Pipeline 23171

**Comment by victorjulien** [comment]
> The header name app-layer-ext.h is a bad name, but I just needed a "clean" header.

`app-layer-ffi.h`? Or a `ffi/app-layer.h` or something?

**Comment by jasonish** [comment]
> > The header name app-layer-ext.h is a bad name, but I just needed a "clean" header.
> 
> `app-layer-ffi.h`? Or a `ffi/app-layer.h` or something?

Not really better names, they imply ffi where these are just normal headers as far as the C code is concerned.  The main point being is I just needed really cleaned up headers to avoid a fight with bindgen to test out bindgen. Should we chose to go to bindgen, we'll probably need some good header cleanup/review for the parts we expose - the big part being not exposing any headers that also depend on headers generated from Rust I think - circular dependencies, thats where things seem to choke.

**Comment by jasonish** [comment]
To see the generated bindings you can do something like:

Build from clean, so only one copy exists:
```
make clean && make
```
Find the bindings:
```
find rust/target -name bindings.rs
```

```
cat rust/target/debug/build/suricata-fbddfaefce7312d9/out/bindings.rs
```

```rust
/* automatically generated by rust-bindgen 0.69.4 */

#[repr(u32)]
#[derive(Debug, Copy, Clone, Hash, PartialEq, Eq)]
pub enum AppLayerEventType {
    APP_LAYER_EVENT_TYPE_TRANSACTION = 1,
    APP_LAYER_EVENT_TYPE_PACKET = 2,
}
pub type SCAppLayerStateGetProgressFn = ::std::option::Option<
    unsafe extern "C" fn(
        alstate: *mut ::std::os::raw::c_void,
        direction: u8,
    ) -> ::std::os::raw::c_int,
>;
pub type SCAppLayerStateGetEventInfoFn = ::std::option::Option<
    unsafe extern "C" fn(
        event_name: *const ::std::os::raw::c_char,
        event_id: *mut ::std::os::raw::c_int,
        event_type: *mut AppLayerEventType,
    ) -> ::std::os::raw::c_int,
>;
pub type SCAppLayerStateGetEventInfoByIdFn = ::std::option::Option<
    unsafe extern "C" fn(
        event_id: ::std::os::raw::c_int,
        event_name: *mut *const ::std::os::raw::c_char,
        event_type: *mut AppLayerEventType,
    ) -> ::std::os::raw::c_int,
>;
```

**Comment by suricata-qa** [comment]
ERROR:
 

ERROR: QA failed on build_fetch.
  

Pipeline 23178

**Pull request closed without merging**